### PR TITLE
Simplified handling of GPU core dumps

### PR DESCRIPTION
**Pull request by jlowe**
Adds configs for simplified GPU core dump handling.  This approach creates a named pipe and listens to the pipe with a background thread to copy the GPU core file to a path that is built with the configured coredump URI prefix.  Coredumps can be configured as lightweight or full, and are by default compressed with zstd before uploading to the (potentially remote) filesystem.

### CODE REVIEW

**Review by revans2 — APPROVED** [review]

**Comment by jlowe** [comment]
build

File: dist/unshimmed-common-from-spark311.txt [hunk 1]
```diff
@@ -8,10 +8,11 @@ com/nvidia/spark/RapidsUDF*
 com/nvidia/spark/Retryable*
 com/nvidia/spark/SQLPlugin*
 com/nvidia/spark/rapids/ColumnarRdd*
-com/nvidia/spark/rapids/GpuColumnVectorUtils*
 com/nvidia/spark/rapids/ExplainPlan.class
 com/nvidia/spark/rapids/ExplainPlan$.class
 com/nvidia/spark/rapids/ExplainPlanBase.class
+com/nvidia/spark/rapids/GpuColumnVectorUtils*
+com/nvidia/spark/rapids/GpuCoreDumpMsg*
```
Comment (gerashegalov) [inline dist/unshimmed-common-from-spark311.txt:15]: I think GpuCoreDumpMsg albeit tiny should be added to the sql-plugin-api module without adding to the tech debt of the classes maintained in this list. Even though GpuCoreDumpHandler as a whole cannot.

**Review by revans2 — DISMISSED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 2]
```diff
@@ -411,6 +411,42 @@ object RapidsConf {
       .integerConf
       .createWithDefault(2)
 
+  val GPU_COREDUMP_DIR = conf("spark.rapids.gpu.coreDump.dir")
+    .doc("The URI to a directory where a GPU core dump will be created if the GPU encounters " +
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:415]: we may want to explicitly say both local and distributed filesystems supported just so its super clear to users

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 3]
```diff
@@ -411,6 +411,42 @@ object RapidsConf {
       .integerConf
       .createWithDefault(2)
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:413]: we may want to add some dev documentation if we are recommending this be done differently on YARN where log aggregation could handle moving core dump.

**Comment by jlowe** [comment]
build

**Review by revans2 — APPROVED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoreDumpHandler.scala [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        val fsOut = dumpFs.create(dumpPath, perms, false, bufferSize,
+          dumpFs.getDefaultReplication(dumpPath), dumpFs.getDefaultBlockSize(dumpPath), null)
+        val out = closeOnExcept(fsOut) { _ =>
+          codec.map(_.compressedOutputStream(fsOut)).getOrElse(fsOut)
+        }
+        withResource(out) { _ =>
+          IOUtils.copy(in, out)
+        }
+        dumpedPath = Some(dumpPath.toString)
+        pluginCtx.send(GpuCoreDumpMsgCompleted(executorId, dumpedPath.get))
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoreDumpHandler.scala:174]: should we have a message for aborted core dumps wrapping an executor exception for easier diagnostics in the driver log?

**Review by gerashegalov — APPROVED** [review]
LGTM

**Review by tgravescs — APPROVED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoreDumpHandler.scala (hunk 4 above)
Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoreDumpHandler.scala:178]: Added this in b4aeffd5f2bc16c9e73c273eea35d914560b76d7

**Comment by jlowe** [comment]
build

**Review by gerashegalov — APPROVED** [review]

**Review by revans2 — APPROVED** [review]

**Pull request merged**